ocaml/xapi/certificates_sync.ml

```diff
@@ -0,0 +1,144 @@
+        ) ;
+        R.ok ()
+    | false -> (
+        let ca_cert_bundle_path = !Xapi_globs.stunnel_cert_path in
+        let pem_leaf = cert.GP.host_cert in
+        let pkcs8_private_key = cert.GP.private_key in
+        let pem_intermediates = String.concat "\n\n" cert.GP.other_certs in
+        let time = Ptime_clock.now () in
+        let tls_on = is_tls_verification_enabled () in
+        match
+          GV.validating_trust_anchors ~pem_leaf ~pkcs8_private_key
+            ~pem_intermediates ~time ~ca_cert_bundle_path
+        with
+        | Error _ as error_msg when tls_on ->
+            error "Host certificate is invalid but required by TLS" ;
+            error_msg
+        | Ok (host_cert, roots) ->
+            let hashes = List.map X509.Certificate.encode_pem roots in
+            let _ref =
+              Xapi_host.add_certificate_to_db ~__context ~host host_cert hashes
+            in
+            R.ok ()
+        | Error _ as error_msg ->
+            warn "The host certificate is invalid but TLS is off" ;
+            error_msg
+      )
+  with e -> Error (`Msg ("installation of host certificate failed", []))
+
+(** determine if the database is up to date by comparing the fingerprint
+  of xapi-ssl.pem with the entry in the database *)
+let is_unchanged ~__context cert_ref cert =
+  let host_pem = cert.GP.host_cert in
+  let ref_hash = Db.Certificate.get_fingerprint ~__context ~self:cert_ref in
+  let* host_cert =
+    Cstruct.of_string host_pem
+    |> X509.Certificate.decode_pem
+    |> R.reword_error (fun (`Msg msg) ->
+           D.info {|Failed to decode certificate because "%s"|} msg ;
+           `Msg (server_certificate_invalid, []))
+  in
+  let cert_hash =
+    X509.Certificate.fingerprint Mirage_crypto.Hash.(`SHA256) host_cert
+    |> Certificates.pp_hash
+  in
+  Ok (cert_hash = ref_hash)
+
+(** [get_server_cert] loads xapi-ssl.pem from the file system and
+  returns it as a record containing the PEM strings *)
+let get_server_cert () =
+  match G.Pem.parse_file !Xapi_globs.server_cert_path with
+  | Error msg ->
+      Error (`Msg (msg, []))
+  | Ok _ as ok ->
+      ok
+
+let update ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let host_uuid = Helpers.get_localhost_uuid () in
+  let cert_refs = Db.Host.get_certificates ~__context ~self:host in
+  let* cert = get_server_cert () in
+  match cert_refs with
+  | [] ->
+      info "Host %s has no active server certificate" host_uuid ;
+      install ~__context ~host cert
+  | [cert_ref] ->
+      let* unchanged = is_unchanged ~__context cert_ref cert in
+      if unchanged then (
+        info "Active server certificate for host %s is unchanged" host_uuid ;
+        Ok ()
+      ) else (
+        info "Server certificate for host %s changed - updating" host_uuid ;
+        let* () = install ~__context ~host cert in
+        remove_cert ~__context cert_ref
+      )
+  | cert_refs ->
```

Comment:
We could merge this match case with the previous one, which would save a little code:
```
| certs_ref::rest as cert_refs ->
  let* unchanged = rest = [] && is_unchanged ~__context cert_ref cert
```

Reply:
I'd like to keep the cases separate: one is the normal case, the other is where we have more than one cert for the host. This should never happen. We install the new one and remove all old ones.

Reply:
We should probably log that something unexpected happened then